## Part 1: Preprocessing

In [66]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

# additional libraries
from category_encoders import OneHotEncoder
import tensorflow as tf

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [67]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [68]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
y_df.head()

Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [69]:
#Viewing the options for X data:
print(attrition_df.columns.unique())

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')


In [70]:
# Create a list of at least 10 column names to use as X data
# Create X_df using your selected columns
X_df = attrition_df[['Age','YearsSinceLastPromotion','YearsInCurrentRole',
                     'RelationshipSatisfaction','PerformanceRating','OverTime',
                     'JobInvolvement','JobSatisfaction','JobRole',
                     'DistanceFromHome','HourlyRate','MaritalStatus',
                     'Education','WorkLifeBalance']]

# Show the data types for X_df
print(X_df.dtypes)


Age                          int64
YearsSinceLastPromotion      int64
YearsInCurrentRole           int64
RelationshipSatisfaction     int64
PerformanceRating            int64
OverTime                    object
JobInvolvement               int64
JobSatisfaction              int64
JobRole                     object
DistanceFromHome             int64
HourlyRate                   int64
MaritalStatus               object
Education                    int64
WorkLifeBalance              int64
dtype: object


In [71]:
#finding total number of observations to get a sense for what train and test split ought to be

print(attrition_df.shape[0])


1470


In [72]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_df,y_df,test_size=0.2,random_state=42)


In [73]:
#finding the number of unique values found in columns OverTime JobRole and MartialStatus

print('OverTime: ', attrition_df.OverTime.nunique())
print('JobRole: ', attrition_df.JobRole.nunique())
print('MaritalStatus: ', attrition_df.MaritalStatus.nunique())

#ok, it looks fine to just use dummies

OverTime:  2
JobRole:  9
MaritalStatus:  3


In [74]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
X_train = pd.get_dummies(X_train)
X_test = pd.get_dummies(X_test)

X_train.head()

Age  YearsSinceLastPromotion  YearsInCurrentRole  \
1097   24                        0                   1   
727    18                        0                   0   
254    29                        0                   2   
1175   39                        1                   4   
1341   31                        0                   8   

      RelationshipSatisfaction  PerformanceRating  JobInvolvement  \
1097                         2                  3               2   
727                          4                  3               3   
254                          4                  3               3   
1175                         3                  4               3   
1341                         1                  3               3   

      JobSatisfaction  DistanceFromHome  HourlyRate  Education  ...  \
1097                1                21          57          2  ...   
727                 4                 5          73          2  ...   
254                 4                20          45          2  ...   
1175                2                12          66          3  ...   
1341                3                20          89          3  ...   

      JobRole_Laboratory Technician  JobRole_Manager  \
1097                              1                0   
727                               0                0   
254                               0                0   
1175                              0                0   
1341                              1                0   

      JobRole_Manufacturing Director  JobRole_Research Director  \
1097                               0                          0   
727                                0                          0   
254                                0                          0   
1175                               1                          0   
1341                               0                          0   

      JobRole_Research Scientist  JobRole_Sales Executive  \
1097                           0                        0   
727                            1                        0   
254                            0                        1   
1175                           0                        0   
1341                           0                        0   

      JobRole_Sales Representative  MaritalStatus_Divorced  \
1097                             0                       1   
727                              0                       0   
254                              0                       1   
1175                             0                       0   
1341                             0                       1   

      MaritalStatus_Married  MaritalStatus_Single  
1097                      0                     0  
727                       0                     1  
254                       0                     0  
1175                      1                     0  
1341                      0                     0  

[5 rows x 25 columns]

In [75]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler = scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


In [92]:
# Create a OneHotEncoder for the Department column
encoder = OneHotEncoder(cols=['Department'])

# Fit the encoder to the training data
encoder.fit(y_train)

# Create two new variables by applying the encoder
# to the training and testing data
y_train_encoded_dept = encoder.transform(y_train)
y_test_encoded_dept = encoder.transform(y_test)

y_train_encoded_dept = y_train_encoded_dept.drop(columns=["Attrition"])
y_test_encoded_dept = y_test_encoded_dept.drop(columns=["Attrition"])

y_train_encoded_dept

Department_1  Department_2  Department_3
1097             1             0             0
727              1             0             0
254              0             1             0
1175             1             0             0
1341             1             0             0
...            ...           ...           ...
1130             1             0             0
1294             1             0             0
860              1             0             0
1459             1             0             0
1126             0             1             0

[1176 rows x 3 columns]

In [93]:
# Create a OneHotEncoder for the Attrition column
encoder = OneHotEncoder(cols=['Attrition'])

# Fit the encoder to the training data
encoder.fit(y_train)

# Create two new variables by applying the encoder
# to the training and testing data
y_train_encoded_attrition = encoder.transform(y_train)
y_test_encoded_attrition = encoder.transform(y_test)

y_train_encoded_attrition = y_train_encoded_attrition.drop(columns=["Department"])
y_test_encoded_attrition = y_test_encoded_attrition.drop(columns=["Department"])

y_train_encoded_attrition



Attrition_1  Attrition_2
1097            1            0
727             1            0
254             1            0
1175            1            0
1341            1            0
...           ...          ...
1130            1            0
1294            1            0
860             0            1
1459            1            0
1126            1            0

[1176 rows x 2 columns]

## Create, Compile, and Train the Model

In [94]:

# Find the number of columns in the X training data
num_cols = X_train_scaled.shape[1]

# Create the input layer
inputs = layers.Input(shape=(num_cols,), name='input_layer')

# Create at least two shared layers
shared_layer_1 = layers.Dense(256, activation='relu', name='shared_layer_1')(inputs)
shared_layer_2 = layers.Dense(128, activation='relu', name='shared_layer_2')(shared_layer_1)


In [95]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
dept_hidden_layer = layers.Dense(64, activation='relu', name='dept_hidden_layer')(shared_layer_2)

# Create the output layer
dept_output = layers.Dense(3, activation='softmax', name='dept_output')(dept_hidden_layer)


In [96]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden_layer = layers.Dense(64, activation='relu', name='attrition_hidden_layer')(shared_layer_2)

# Create the output layer
attrition_output = layers.Dense(2, activation='softmax', name='attrition_output')(attrition_hidden_layer)


In [97]:
# Create the model
model = Model(inputs=inputs, outputs=[dept_output, attrition_output], name='attrition_model')

# Compile the model
model.compile(optimizer='adam', loss={'dept_output': 'categorical_crossentropy',
                                    'attrition_output': 'binary_crossentropy'},
              metrics={'dept_output': 'accuracy',
                       'attrition_output': 'accuracy'})

# Summarize the model
model.summary()


Model: "attrition_model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_layer (InputLayer)    [(None, 25)]                 0         []                            
                                                                                                  
 shared_layer_1 (Dense)      (None, 256)                  6656      ['input_layer[0][0]']         
                                                                                                  
 shared_layer_2 (Dense)      (None, 128)                  32896     ['shared_layer_1[0][0]']      
                                                                                                  
 dept_hidden_layer (Dense)   (None, 64)                   8256      ['shared_layer_2[0][0]']      
                                                                                    

In [102]:
# Need to set run_functions_eagerly to true in order for my model to run. I'm not sure where exactly the issue was, but the fit seems good.

tf.config.run_functions_eagerly(True)
# Train the model
model.fit(X_train_scaled,
          {'dept_output': y_train_encoded_dept,
           'attrition_output': y_train_encoded_attrition},
          epochs=100,
          validation_data=(X_test_scaled,
                           {'dept_output': y_test_encoded_dept,
                            'attrition_output': y_test_encoded_attrition}))


Epoch 1/100


/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/structured_function.py:258: UserWarning: Even though the `tf.config.experimental_run_functions_eagerly` option is set, this option does not apply to tf.data functions. To force eager execution of tf.data functions, please use `tf.data.experimental.enable_debug_mode()`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/structured_function.py:258: UserWarning: Even though the `tf.config.experimental_run_functions_eagerly` option is set, this option does not apply to tf.data functions. To force eager execution of tf.data functions, please use `tf.data.experimental.enable_debug_mode()`.
  warnings.warn(


37/37 [==============================] - ETA: 0s - loss: 1.0920 - dept_output_loss: 0.5788 - attrition_output_loss: 0.5132 - dept_output_accuracy: 0.7730 - attrition_output_accuracy: 0.8078

/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/structured_function.py:258: UserWarning: Even though the `tf.config.experimental_run_functions_eagerly` option is set, this option does not apply to tf.data functions. To force eager execution of tf.data functions, please use `tf.data.experimental.enable_debug_mode()`.
  warnings.warn(


37/37 [==============================] - 5s 122ms/step - loss: 1.0920 - dept_output_loss: 0.5788 - attrition_output_loss: 0.5132 - dept_output_accuracy: 0.7730 - attrition_output_accuracy: 0.8078 - val_loss: 0.5873 - val_dept_output_loss: 0.1998 - val_attrition_output_loss: 0.3875 - val_dept_output_accuracy: 0.9694 - val_attrition_output_accuracy: 0.8673
Epoch 2/100
37/37 [==============================] - 4s 98ms/step - loss: 0.4915 - dept_output_loss: 0.1056 - attrition_output_loss: 0.3859 - dept_output_accuracy: 0.9583 - attrition_output_accuracy: 0.8342 - val_loss: 0.4644 - val_dept_output_loss: 0.0906 - val_attrition_output_loss: 0.3738 - val_dept_output_accuracy: 0.9626 - val_attrition_output_accuracy: 0.8741
Epoch 3/100
37/37 [==============================] - 3s 71ms/step - loss: 0.4054 - dept_output_loss: 0.0623 - attrition_output_loss: 0.3431 - dept_output_accuracy: 0.9736 - attrition_output_accuracy: 0.8546 - val_loss: 0.4648 - val_dept_output_loss: 0.0854 - val_attrition_ou

In [103]:
# prompt: # Evaluate the model with the testing data

# Evaluate the model with the testing data
model.evaluate(X_test_scaled,
              {'dept_output': y_test_encoded_dept,
               'attrition_output': y_test_encoded_attrition})


 4/10 [===========>..................] - ETA: 0s - loss: 2.7978 - dept_output_loss: 0.2548 - attrition_output_loss: 2.5430 - dept_output_accuracy: 0.9766 - attrition_output_accuracy: 0.8281

/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/structured_function.py:258: UserWarning: Even though the `tf.config.experimental_run_functions_eagerly` option is set, this option does not apply to tf.data functions. To force eager execution of tf.data functions, please use `tf.data.experimental.enable_debug_mode()`.
  warnings.warn(


10/10 [==============================] - 0s 24ms/step - loss: 2.1704 - dept_output_loss: 0.3509 - attrition_output_loss: 1.8194 - dept_output_accuracy: 0.9694 - attrition_output_accuracy: 0.8401


[2.170372486114502,
 0.35094770789146423,
 1.8194249868392944,
 0.9693877696990967,
 0.8401360511779785]

In [116]:
# prompt: could you generate a confusion matrix of sort on the output of this data?

# Generate predictions for the testing data
y_pred = model.predict(X_test_scaled)

# Extract the predicted classes for the Department and Attrition outputs
dept_pred = np.argmax(y_pred[0], axis=1)
attrition_pred = np.argmax(y_pred[1], axis=1)

# Extract the true classes for the Department and Attrition outputs
dept_true = np.argmax(y_test_encoded_dept.values, axis=1)
attrition_true = np.argmax(y_test_encoded_attrition.values, axis=1)

# Create confusion matrices for the Department and Attrition outputs
from sklearn.metrics import confusion_matrix

dept_cm = confusion_matrix(dept_true, dept_pred)
attrition_cm = confusion_matrix(attrition_true, attrition_pred)

# Print the confusion matrices
print("Department Confusion Matrix:")
print(dept_cm)
print("\nAttrition Confusion Matrix:")
print(attrition_cm)


10/10 [==============================] - 0s 20ms/step
Department Confusion Matrix:
[[191   5   0]
 [  1  84   0]
 [  2   1  10]]

Attrition Confusion Matrix:
[[234  21]
 [ 26  13]]


**Interpretation of Confusion matrix written by AI**

`Department:`

Cell (0,0): This represents the true positives (TP) for Department A. The model correctly predicted 191 instances as Department A.
Cell (0,1): This represents the false positives (FP) for Department A. The model incorrectly predicted 5 instances as Department B.
Cell (0,2): This represents the false negatives (FN) for Department A. The model missed 2 instances of Department A.
Cell (1,0): This represents the false negatives (FN) for Department B. The model missed 1 instance of Department B.
Cell (1,1): This represents the true positives (TP) for Department B. The model correctly predicted 84 instances as Department B.
Cell (1,2): This represents the false positives (FP) for Department B. The model incorrectly predicted 1 instance as Department C.
Cell (2,0): This represents the false positives (FP) for Department C. The model incorrectly predicted 2 instances as Department A.
Cell (2,1): This represents the false negatives (FN) for Department C. The model missed 1 instance of Department C.
Cell (2,2): This represents the true positives (TP) for Department C. The model correctly predicted 10 instances as Department C.


`Attrition:`

Cell (0,0): This represents the true negatives (TN) for the “No Attrition” class. The model correctly predicted 234 instances as “No Attrition.”
Cell (0,1): This represents the false positives (FP) for the “No Attrition” class. The model incorrectly predicted 21 instances as “Attrition” when they were actually “No Attrition.”
Cell (1,0): This represents the false negatives (FN) for the “Attrition” class. The model missed 26 instances of actual “Attrition.”
Cell (1,1): This represents the true positives (TP) for the “Attrition” class. The model correctly predicted 13 instances as “Attrition.”

In [120]:
# Print the accuracy for both department and attrition

print('Department Accuracy:', model.evaluate(X_test_scaled,
              {'dept_output': y_test_encoded_dept,
               'attrition_output': y_test_encoded_attrition})[3])
print('Attrition Accuracy:', model.evaluate(X_test_scaled,
              {'dept_output': y_test_encoded_dept,
               'attrition_output': y_test_encoded_attrition})[4])


 2/10 [=====>........................] - ETA: 0s - loss: 2.6052 - dept_output_loss: 0.1448 - attrition_output_loss: 2.4604 - dept_output_accuracy: 0.9844 - attrition_output_accuracy: 0.7969

/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/structured_function.py:258: UserWarning: Even though the `tf.config.experimental_run_functions_eagerly` option is set, this option does not apply to tf.data functions. To force eager execution of tf.data functions, please use `tf.data.experimental.enable_debug_mode()`.
  warnings.warn(


10/10 [==============================] - 1s 58ms/step - loss: 2.1704 - dept_output_loss: 0.3509 - attrition_output_loss: 1.8194 - dept_output_accuracy: 0.9694 - attrition_output_accuracy: 0.8401
Department Accuracy: 0.9693877696990967
 1/10 [==>...........................] - ETA: 1s - loss: 2.4296 - dept_output_loss: 0.2896 - attrition_output_loss: 2.1400 - dept_output_accuracy: 0.9688 - attrition_output_accuracy: 0.8125

/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/structured_function.py:258: UserWarning: Even though the `tf.config.experimental_run_functions_eagerly` option is set, this option does not apply to tf.data functions. To force eager execution of tf.data functions, please use `tf.data.experimental.enable_debug_mode()`.
  warnings.warn(


10/10 [==============================] - 1s 87ms/step - loss: 2.1704 - dept_output_loss: 0.3509 - attrition_output_loss: 1.8194 - dept_output_accuracy: 0.9694 - attrition_output_accuracy: 0.8401
Attrition Accuracy: 0.8401360511779785


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?


YOUR ANSWERS HERE

1. 
It appears that the issue lies with the unbalanced nature of the data rather than the model's overall accuracy. While the model may exhibit high accuracy, a deeper analysis of the confusion matrix reveals shortcomings in accurately predicting employee attrition. Despite the seemingly good performance, the model struggles notably with detecting instances of attrition, indicating a need for attention to the imbalance within the dataset to improve its effectiveness in this specific area.
2. I opted for softmax activation function assuming its effectiveness in addressing multiclass problems. However, I overlooked the potential benefits of using a sigmoid function, which might have yielded more valuable results.
3. I'm not entirely certain, but I believe incorporating additional features into the model could have been beneficial. Furthermore, tweaking the model's hyperparameters, such as experimenting with different activation functions, might help enhance its performance through hyperparameter tuning.